### PR TITLE
Migrate renorm to ATen (CPU and CUDA)

### LINKED ISSUE #24616: Migrate `renorm` and `renorm_` from the TH to Aten (CUDA)
Porting TH operators is essential for code simplicity and performance reasons.

Porting guides and Q&A are available in umbrella issue: #24507

Feel free to add @VitalyFedyunin as a reviewer to get a prioritized review.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser

### LINKED ISSUE #24754: Migrate `renorm` and `renorm_` from the TH to Aten (CPU)
Porting TH operators is essential for code simplicity and performance reasons.

Porting guides and Q&A are available in umbrella issue: #24507

Feel free to add @VitalyFedyunin as a reviewer to get a prioritized review.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser

### PR DESCRIPTION
**Pull request by peterbell10**
Resubmit of #59108, closes #24754, closes #24616

This reuses `linalg_vector_norm` to calculate the norms. I just add a new kernel that turns  the norm into a normalization factor, then multiply the original tensor using a normal broadcasted `mul` operator. The result is less code, and better performance to boot.

#### Benchmarks (CPU):
|     Shape    | Dim |  Before | After (1 thread) | After (8 threads) |
|:------------:|:---:|--------:|-----------------:|------------------:|
| (10, 10, 10) | 0   | 11.6 us |           4.2 us |            4.2 us |
|              | 1   | 14.3 us |           5.2 us |            5.2 us |
|              | 2   | 12.7 us |           4.6 us |            4.6 us |
| (50, 50, 50) | 0   |  330 us |           120 us |           24.4 us |
|              | 1   |  350 us |           135 us |           28.2 us |
|              | 2   |  417 us |           130 us |           24.4 us |

#### Benchmarks (CUDA)
|     Shape    | Dim |  Before |   After |
|:------------:|:---:|--------:|--------:|
| (10, 10, 10) | 0   | 12.5 us | 12.1 us |
|              | 1   | 13.1 us | 12.2 us |
|              | 2   | 13.1 us | 11.8 us |
| (50, 50, 50) | 0   | 33.7 us | 11.6 us |
|              | 1   | 36.5 us | 15.8 us |
|              | 2   | 41.1 us |   15 us |

### CODE REVIEW

File: aten/src/ATen/native/Normalization.cpp [hunk 1]
```diff
@@ -615,4 +635,39 @@ std::tuple<Tensor, Tensor, Tensor> batch_norm_backward_cpu(const Tensor& grad_ou
...
+  auto dtype = self.scalar_type();
+  auto acc_type = at::toAccumulateType(dtype, /*is_cuda=*/true);
+  Tensor norm;
+  if (acc_type != dtype) {
+    norm = at::linalg_vector_norm(self, p.toDouble(), reduce_dims,
+                                  /*keepdim=*/true, /*dtype=*/acc_type);
+  } else {
+    norm = at::linalg_vector_norm(self, p.toDouble(), reduce_dims,
+                                  /*keepdim=*/true);
+  }
```
Comment (peterbell10) [inline aten/src/ATen/native/Normalization.cpp:658]: For complex types, the input is cast to `dtype` and the output has type `dtype` so it has to be complex. However, if I leave it out entirely, the result is real. Seems like an odd way for this to work.

Comment (ngimel) [reply aten/src/ATen/native/Normalization.cpp:658]: cc @kurtamohler for norm `dtype` behavior. Should we disallow specifying non-complex `dtype` for complex? I believe there was a warning about that at some point. But then, it doesn't allow users to specify different precision if the input is complex.

Comment (kurtamohler) [reply aten/src/ATen/native/Normalization.cpp:658]: I agree that it's odd behavior. The explanation is that the input is converted to `dtype` before the norm is calculated, so if we allowed a non-complex `dtype` arg when the input is complex, we would be discarding the imaginary part. So at the moment we throw an error in that case: https://github.com/pytorch/pytorch/blob/afdfd2288ab/aten/src/ATen/native/LinearAlgebra.cpp#L2271

Comment (peterbell10) [reply aten/src/ATen/native/Normalization.cpp:658]: If the result was always real then `norm(tensor, ..., dtype=dtype)` would behaave just like `norm(tensor.to(dtype), ...)`. I think that's reasonable.

**Review by ngimel — APPROVED** [review]

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 2]
```diff
@@ -784,30 +785,32 @@ Tensor sparse_sparse_matmul_backward(
...
+
+  Tensor norm;
+  if (acc_type != dtype) {
+    norm = at::linalg_vector_norm(
+        self, p, reduce_dims, /*keepdim=*/true, /*dtype=*/acc_type);
+  } else {
+    norm = at::linalg_vector_norm(
+        self, p, reduce_dims, /*keepdim=*/true);
+  }
+  auto grad_output = (self.conj() * grad).sum(
```
Comment (ngimel) [inline torch/csrc/autograd/FunctionsManual.cpp:806]: out of curiosity, is discarding imaginary part expected here?

Comment (peterbell10) [reply torch/csrc/autograd/FunctionsManual.cpp:806]: Yes. The norm is real-valued so it's gradient should be as well. In fact the gradcheck tests fail if I don't intentionally cast to real.

Comment (ngimel) [reply torch/csrc/autograd/FunctionsManual.cpp:806]: But `/*dtype=*/c10::toValueType(acc_type)` will always be throwing warnings when input is converted to real, even though that's what we want. Maybe `(self.conj() * grad).real().contiguous().sum(...)`? (`.contiguous` is optional, on cpu I suspect discontiguous reduction will be very bad, on the gpu on the other hand .contiguous() + reduction will be slower than just reduction, and in any case, people don't seem to care about renorm backward performance).

Comment (peterbell10) [reply torch/csrc/autograd/FunctionsManual.cpp]: > Maybe `(self.conj() * grad).real().contiguous().sum(...)`?

The awkward thing is that `at::real` isn't a no-op for real types. It throws an error so you can't write a one-liner. But if this way creates a warning then I guess there's no ignoring it.

**Pull request closed without merging**